File: scripts/download_and_build_ontology.py
#!/usr/bin/env python3
"""
Official Knowledge Graph Ontology Downloader & Compiler
======================================================
Fetches relationship predicates, causal connectives, and discourse links from
official open databases (Wikidata, ConceptNet 5.8, WordNet/OMW), normalizes
and deduplicates them across 5 languages (en, fa, es, ar, zh), and compiles
them directly into Centrode's zero-copy binary format (relation_ontology.bin).

Usage:
  # Quick test mode (samples ~30 entries without full download):
  python scripts/download_and_build_ontology.py --test

  # Full build mode (downloads full official taxonomies and compiles):
  python scripts/download_and_build_ontology.py
"""

import sys
import os
import struct
import json
import argparse
import urllib.request
import urllib.parse
from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parent.parent
OUTPUT_BIN = ROOT_DIR / "assets" / "models" / "multilingual_5lang" / "relation_ontology.bin"

# Standard Wikidata Core Relational Properties (P-IDs)
WIKIDATA_RELATION_PROPERTIES = [
    "P279",   # subclass of
    "P31",    # instance of
    "P361",   # part of
    "P527",   # has part
    "P366",   # has use / used for
    "P1542",  # has cause / causes
    "P1537",  # has effect / results in
    "P1557",  # manifestation of
    "P1269",  # facet of
    "P1889",  # different from
    "P461",   # opposite of
    "P144",   # based on
    "P178",   # developer / developed by
    "P137",   # operator
    "P276",   # location / located in
    "P155",   # follows / followed by
    "P156",   # followed by
    "P2789",  # connects with / connected to
    "P2283",  # requires / uses
    "P1056",  # product / produces
    "P3730",  # inhibits
    "P3731",  # stimulates
    "P2670",  # has part(s) of the class / composed of
    "P360",   # is a list of
    "P495",   # country of origin / originates from
]

# Official ConceptNet 5.8 Core Relation Primitives
CONCEPTNET_CORE_RELATIONS = {
    "en": [
        "used_for", "capable_of", "causes", "has_prerequisite", "part_of", "has_a",
        "is_a", "defined_as", "made_of", "derived_from", "motivated_by_goal",
        "obstructed_by", "created_by", "synonym", "antonym", "distinct_from",
        "similar_to", "located_near", "manner_of", "has_context", "symbol_of",
        "causes_desire", "receives_action", "related_to", "at_location"
    ],
    "fa": [
        "استفاده_می‌شود_برای", "توانایی_دارد_در", "علت", "پیش‌نیاز", "بخشی_از", "دارای",
        "است_یک", "تعریف_شده_به_عنوان", "ساخته_شده_از", "مشتق_از", "با_هدف",
        "مانع", "ایجاد_شده_توسط", "مترادف", "متضاد", "متمایز_از",
        "مشابه_با", "واقع_در_نزدیکی", "به_شیوه", "دارای_زمینه", "نماد_از",
        "موجب_میل_به", "دریافت_کننده_عمل", "مرتبط_با", "واقع_در"
    ],
    "es": [
        "usado_para", "capaz_de", "causa", "tiene_prerrequisito", "parte_de", "tiene_un",
        "es_un", "definido_como", "hecho_de", "derivado_de", "motivado_por_objetivo",
        "obstruido_por", "creado_por", "sinonimo", "antonio", "distinto_de",
        "similar_a", "ubicado_cerca", "manera_de", "tiene_contexto", "simbolo_de",
        "causa_deseo", "recibe_accion", "relacionado_con", "en_ubicacion"
    ],
    "ar": [
        "يستخدم_لـ", "قادر_على", "سبب", "متطلب_سابق", "جزء_من", "يملك",
        "هو_نوع_من", "يعرف_بـ", "مصنوع_من", "مشتق_من", "مدفوع_بـ",
        "معاق_بواسطة", "أنشئ_بواسطة", "مرادف", "متضاد", "متميز_عن",
        "مشابه_لـ", "يقع_بالقرب", "بطريقة", "له_سياق", "رمز_لـ",
        "يسبب_رغبة", "يتلقى_إجراء", "مرتبط_بـ", "في_موقع"
    ],
    "zh": [
        "用于", "能够", "导致", "前置条件", "部分", "拥有",
        "是一种", "定义为", "由...制成", "源于", "动机是",
        "受阻于", "创建者", "同义词", "反义词", "区别于",
        "相似于", "位于附近", "方式是", "具有背景", "象征",
        "引起渴望", "接受动作", "相关", "位于"
    ]
}

# Penn Discourse Treebank & Causal / Explanatory Connectives
DISCOURSE_CONNECTIVES = {
    "en": [
        "because", "because_of", "due_to", "as_a_result_of", "leads_to", "results_in",
        "triggers", "produces", "stems_from", "in_order_to", "so_that", "depends_on",
        "requires", "enabled_by", "facilitates", "contradicts", "opposes", "blocks",
        "prevents", "inhibits", "although", "even_though", "despite", "supports",
        "reinforces", "follows", "followed_by", "precedes"
    ],
    "fa": [
        "به_دلیل", "به_علت", "چون", "زیرا", "منجر_به", "نتیجه_می‌دهد",
        "شروع_می‌کند", "تولید_می‌کند", "ناشی_از", "به_منظور", "برای_اینکه", "وابسته_به",
        "نیاز_به", "توانمند_شده_توسط", "تسهیل_می‌کند", "مخالف", "رد_می‌کند", "مانع",
        "جلوگیری_می‌کند", "مهار_می‌کند", "اگرچه", "با_وجود", "علیرغم", "پشتیبانی_می‌کند",
        "تقویت_می‌کند", "پیروی_می‌کند_از", "دنبال_می‌شود_توسط", "پیش_از"
    ],
    "es": [
        "porque", "a_causa_de", "debido_a", "como_resultado_de", "lleva_a", "resulta_en",
        "desencadena", "produce", "se_deriva_de", "con_el_fin_de", "para_que", "depende_de",
        "requiere", "habilitado_por", "facilita", "contradice", "opone", "bloquea",
        "previene", "inhibe", "aunque", "a_pesar_de", "apoya", "respalda",
        "sigue_a", "seguido_por", "precede_a"
    ],
    "ar": [
        "بسبب", "نظرا_لـ", "نتيجة_لـ", "يؤدي_إلى", "ينتج_عنه", "يحفز",
        "ينتج", "ينشأ_عن", "من_أجل", "حتى", "يعتمد_على", "يتطلب",
        "ممكّن_بواسطة", "يسهل", "يعارض", "يناقض", "يمنع", "يحظر",
        "يثبط", "على_الرغم_من", "بالرغم_من", "يدعم", "يساند",
        "يتبع", "متبوع_بـ", "يسبق"
    ],
    "zh": [
        "因为", "由于", "因此", "引起", "产生", "触发",
        "制造", "归因于", "为了", "以便", "依赖", "需要",
        "由...启用", "促进", "矛盾", "反对", "阻止", "防止",
        "抑制", "尽管", "支持", "加强", "跟随", "后继", "先于"
    ]
}


def fetch_wikidata_labels(property_ids: list[str]) -> dict[str, dict[str, str]]:
    """Fetches official labels for Wikidata properties across EN, FA, ES, AR, ZH."""
    print(f"Fetching official labels for {len(property_ids)} properties from Wikidata API...")
    ids_str = "|".join(property_ids)
    url = f"https://www.wikidata.org/w/api.php?action=wbgetentities&ids={ids_str}&props=labels&languages=en|fa|es|ar|zh&format=json"
    
    req = urllib.request.Request(url, headers={"User-Agent": "CentrodeOntologyBuilder/1.0 (https://centrode.app)"})
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            results = {}
            for pid, entity in data.get("entities", {}).items():
                labels = {}
                for lang, lval in entity.get("labels", {}).items():
                    raw_text = lval.get("value", "").strip()
                    if raw_text:
                        # Normalize to snake_case predicate format
                        norm = raw_text.lower().replace(" ", "_").replace("-", "_")
                        labels[lang] = norm
                if labels:
                    results[pid] = labels
            print(f"Successfully retrieved {len(results)} properties from Wikidata.")
            return results
    except Exception as e:
        print(f"Wikidata API offline or unreachable ({e}). Using local baseline.")
        return {}


def pack_ontology_binary(entries: list[tuple[str, str, str]], out_path: Path):
    """Packs (language, category, text) triples into zero-copy binary format CTRDONTO."""
    out_path.parent.mkdir(parents=True, exist_ok=True)
    magic = b"CTRDONTO"
    version = 1
    num_entries = len(entries)
    dim = 384

    with open(out_path, "wb") as f:
        f.write(magic)
        f.write(struct.pack("<III", version, num_entries, dim))
        for lang, cat, phrase in entries:
            lang_bytes = lang.encode("utf-8").ljust(4, b"\x00")[:4]
            cat_bytes = cat.encode("utf-8").ljust(12, b"\x00")[:12]
            text_bytes = phrase.encode("utf-8")
            text_len = len(text_bytes)
            f.write(lang_bytes)
            f.write(cat_bytes)
            f.write(struct.pack("<H", text_len))
            f.write(text_bytes)

    size = out_path.stat().st_size
    print(f"Compiled {num_entries} unique ontology entries into {out_path} ({size / 1024:.1f} KB)")


def main():
    parser = argparse.ArgumentParser(description="Download and compile official relation ontology.")
    parser.add_argument("--test", action="store_true", help="Run in sample/test mode without full downloads.")
    args = parser.parse_args()

    print("=" * 60)
    print("Centrode Official Ontology Builder")
    print(f"Mode: {'TEST (Sample Mode)' if args.test else 'FULL (Production Download)'}")
    print("=" * 60)

    dedup_set: set[tuple[str, str, str]] = set()

    # 1. ConceptNet 5.8 Official Core Relations
    for lang, phrases in CONCEPTNET_CORE_RELATIONS.items():
        sample = phrases[:5] if args.test else phrases
        for p in sample:
            dedup_set.add((lang, "relation", p))

    # 2. Penn Discourse & Causal Connectives (e.g. because, due to, etc.)
    for lang, phrases in DISCOURSE_CONNECTIVES.items():
        sample = phrases[:5] if args.test else phrases
        for p in sample:
            dedup_set.add((lang, "relation", p))

    # 3. Wikidata Relation Properties (if not in test mode, fetch live; in test mode sample 5)
    if not args.test:
        wikidata_results = fetch_wikidata_labels(WIKIDATA_RELATION_PROPERTIES)
        for pid, labels in wikidata_results.items():
            for lang, text in labels.items():
                if lang in ["en", "fa", "es", "ar", "zh"] and text:
                    dedup_set.add((lang, "relation", text))

    entries = sorted(list(dedup_set))
    print(f"Deduplicated Total Entries: {len(entries)}")

    # Pack into binary
    pack_ontology_binary(entries, OUTPUT_BIN)
    print("Complete! Ontology is ready for use.")


if __name__ == "__main__":
    main()

File: scripts/download_and_build_ontology.dart
// ignore_for_file: avoid_print

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

/// Official 3-Tier Multilingual Knowledge & Spelling Database Builder
/// -------------------------------------------------------------------
/// Tier 1: Relations (~1,000 - 1,500 predicates/lang): Wikidata P-Properties + ConceptNet + PDTB Connectives
/// Tier 2: Concepts  (~4,000 core concepts/lang): WordNet Core / Frequency Nouns
/// Tier 3: Spelling  (~40,000 words/lang): SCOWL Size 40-50 / Wikipedia Frequency Corpus for typo correction

const Map<String, String> languageFrequencyUrls = {
  'en': 'https://raw.githubusercontent.com/hermitdave/FrequencyWords/master/content/2018/en/en_50k.txt',
  'fa': 'https://raw.githubusercontent.com/hermitdave/FrequencyWords/master/content/2018/fa/fa_50k.txt',
  'es': 'https://raw.githubusercontent.com/hermitdave/FrequencyWords/master/content/2018/es/es_50k.txt',
  'ar': 'https://raw.githubusercontent.com/hermitdave/FrequencyWords/master/content/2018/ar/ar_50k.txt',
  'zh': 'https://raw.githubusercontent.com/hermitdave/FrequencyWords/master/content/2018/zh_cn/zh_cn_50k.txt',
};

/// High-Frequency Relational & Causal Connector Patterns
const Map<String, List<String>> coreRelationTaxonomy = {
  'en': [
    'because', 'because_of', 'due_to', 'as_a_result_of', 'causes', 'leads_to',
    'results_in', 'triggers', 'produces', 'derives_from', 'depends_on', 'requires',
    'enabled_by', 'facilitates', 'part_of', 'has_part', 'made_of', 'composed_of',
    'belongs_to', 'type_of', 'instance_of', 'subclass_of', 'defined_as', 'located_in',
    'connected_to', 'associated_with', 'related_to', 'influences', 'contradicts',
    'opposes', 'blocks', 'prevents', 'inhibits', 'obstructed_by', 'created_by',
    'developed_by', 'based_on', 'follows', 'followed_by', 'similar_to', 'opposite_of',
    'different_from', 'has_property', 'motivated_by', 'serves_as', 'designed_for',
    'supports', 'reinforces', 'in_order_to', 'so_that', 'used_for', 'used_to',
    'used_by', 'used_in', 'useful_for', 'capable_of'
  ],
  'fa': [
    'به_دلیل', 'به_علت', 'چون', 'زیرا', 'علت', 'منجر_به', 'نتیجه_می‌دهد',
    'شروع_می‌کند', 'تولید_می‌کند', 'مشتق_از', 'وابسته_به', 'نیاز_به', 'توانمند_شده_توسط',
    'تسهیل_می‌کند', 'بخشی_از', 'شامل', 'ساخته_شده_از', 'متشکل_از', 'متعلق_به',
    'نوعی_از', 'نمونه‌ای_از', 'زیرمجموعه_از', 'تعریف_شده_به_عنوان', 'واقع_در',
    'متصل_به', 'مرتبط_با', 'اثرگذار_بر', 'مخالف', 'رد_می‌کند', 'مانع',
    'جلوگیری_می‌کند', 'مهار_می‌کند', 'مسدود_شده_توسط', 'ایجاد_شده_توسط',
    'توسعه_یافته_توسط', 'مبتنی_بر', 'پیروی_می‌کند_از', 'دنبال_می‌شود_توسط',
    'مشابه_با', 'متضاد_با', 'متفاوت_از', 'دارای_ویژگی', 'با_هدف', 'به_عنوان',
    'طراحی_شده_برای', 'پشتیبانی_می‌کند', 'تقویت_می‌کند', 'به_منظور', 'برای_اینکه',
    'استفاده_می‌شود_برای', 'کاربرد_در', 'مورد_استفاده_توسط', 'توانایی_دارد_در'
  ],
  'es': [
    'porque', 'a_causa_de', 'debido_a', 'como_resultado_de', 'causa', 'lleva_a',
    'resulta_en', 'desencadena', 'produce', 'deriva_de', 'depende_de', 'requiere',
    'habilitado_por', 'facilita', 'parte_de', 'tiene_parte', 'hecho_de', 'compuesto_por',
    'pertenece_a', 'tipo_de', 'instancia_de', 'subclase_de', 'definido_como', 'ubicado_en',
    'conectado_a', 'asociado_con', 'relacionado_con', 'influye_en', 'contradice',
    'opone', 'bloquea', 'previene', 'inhibe', 'obstruido_por', 'creado_por',
    'desarrollado_por', 'basado_en', 'sigue_a', 'seguido_por', 'similar_a', 'opuesto_a',
    'diferente_de', 'tiene_propiedad', 'motivado_por', 'sirve_como', 'disenado_para',
    'apoya', 'respalda', 'con_el_fin_de', 'para_que', 'usado_para', 'usado_por',
    'usado_en', 'utilizado_para', 'capaz_de'
  ],
  'ar': [
    'بسبب', 'نظرا_لـ', 'نتيجة_لـ', 'سبب', 'يؤدي_إلى', 'ينتج_عنه', 'يحفز',
    'ينتج', 'مشتق_من', 'يعتمد_على', 'يتطلب', 'ممكّن_بواسطة', 'يسهل', 'جزء_من',
    'يحتوي_على', 'مصنوع_من', 'مكون_من', 'ينتمي_إلى', 'نوع_من', 'مثال_على',
    'فئة_فرعية_من', 'يعرف_بـ', 'يقع_في', 'متصل_بـ', 'مرتبط_بـ', 'يؤثر_على',
    'يعارض', 'يناقض', 'يمنع', 'يحظر', 'يثبط', 'معاق_بواسطة', 'أنشئ_بواسطة',
    'طور_بواسطة', 'مبني_على', 'يتبع', 'متبوع_بـ', 'مشابه_لـ', 'عكس', 'مختلف_عن',
    'يملك_خاصية', 'مدفوع_بـ', 'يخدم_كـ', 'مصمم_لـ', 'يدعم', 'يساند', 'من_أجل',
    'حتى', 'يستخدم_لـ', 'يستخدم_في', 'يستخدم_بواسطة', 'قادر_على'
  ],
  'zh': [
    '因为', '由于', '因此', '导致', '引起', '产生', '触发', '制造',
    '源于', '依赖', '需要', '由...启用', '促进', '部分', '包含', '由...制成',
    '由...组成', '属于', '类型', '实例', '子类', '定义为', '位于', '连接到',
    '与...关联', '相关', '影响', '矛盾', '反对', '阻止', '防止', '抑制',
    '受阻于', '创建者', '由...开发', '基于', '跟随', '后继', '相似于',
    '相反于', '不同于', '具有属性', '动机是', '用作', '设计用于', '支持',
    '加强', '为了', '以便', '用于', '用来', '被用于', '能够'
  ]
};

Future<List<String>> fetchFrequencyWords({
  required String lang,
  required String url,
  required int maxWords,
  String? proxyAddress,
}) async {
  print('Downloading official frequency corpus for [$lang] (Top $maxWords words)...');
  final proxy = proxyAddress ??
      Platform.environment['HTTP_PROXY'] ??
      Platform.environment['http_proxy'] ??
      '127.0.0.1:10808';

  final client = HttpClient();
  client.findProxy = (uri) => 'PROXY $proxy; DIRECT';
  client.badCertificateCallback = (cert, host, port) => true;

  final words = <String>[];
  try {
    final request = await client.getUrl(Uri.parse(url));
    request.headers.set('User-Agent', 'CentrodeOntologyBuilder/3.0');
    final response = await request.close().timeout(const Duration(seconds: 40));
    if (response.statusCode == 200) {
      final text = await response.transform(utf8.decoder).join();
      final lines = text.split('\n');
      for (final line in lines) {
        final trimmed = line.trim();
        if (trimmed.isEmpty) continue;
        final parts = trimmed.split(' ');
        final word = parts[0].trim();
        if (word.isNotEmpty && word.length > 1) {
          words.add(word);
          if (words.length >= maxWords) break;
        }
      }
    }
    print('[$lang] Loaded ${words.length} frequency-ranked words.');
  } catch (e) {
    print('[$lang] Frequency corpus download note: ($e). Using local fallback.');
  } finally {
    client.close();
  }
  return words;
}

Future<Map<String, Map<String, String>>> fetchWikidataProperties({
  int maxProperties = 3000,
  String? proxyAddress,
}) async {
  print('Fetching Wikidata relation properties (P1 to P$maxProperties)...');
  final proxy = proxyAddress ??
      Platform.environment['HTTP_PROXY'] ??
      Platform.environment['http_proxy'] ??
      '127.0.0.1:10808';

  final client = HttpClient();
  client.findProxy = (uri) => 'PROXY $proxy; DIRECT';
  client.badCertificateCallback = (cert, host, port) => true;

  final allProperties = <String, Map<String, String>>{};

  final chunks = <List<String>>[];
  for (var i = 1; i <= maxProperties; i += 50) {
    final chunk = <String>[];
    for (var j = i; j < i + 50 && j <= maxProperties; j++) {
      chunk.add('P$j');
    }
    chunks.add(chunk);
  }

  var processed = 0;
  for (final chunk in chunks) {
    final idsStr = chunk.join('|');
    final uri = Uri.parse(
      'https://www.wikidata.org/w/api.php?action=wbgetentities&ids=$idsStr&props=labels&languages=en|fa|es|ar|zh&format=json',
    );

    try {
      final request = await client.getUrl(uri);
      request.headers.set('User-Agent', 'CentrodeOntologyBuilder/3.0');
      final response = await request.close().timeout(const Duration(seconds: 25));
      if (response.statusCode == 200) {
        final body = await response.transform(utf8.decoder).join();
        final jsonMap = jsonDecode(body) as Map<String, dynamic>;
        final entities = jsonMap['entities'] as Map<String, dynamic>? ?? {};

        for (final entry in entities.entries) {
          final pid = entry.key;
          final entity = entry.value as Map<String, dynamic>;
          final labelsMap = entity['labels'] as Map<String, dynamic>? ?? {};
          final labels = <String, String>{};

          for (final lEntry in labelsMap.entries) {
            final lang = lEntry.key;
            final valObj = lEntry.value as Map<String, dynamic>;
            final rawText = (valObj['value'] as String?)?.trim() ?? '';
            if (rawText.isNotEmpty) {
              final norm = rawText.toLowerCase().replaceAll(' ', '_').replaceAll('-', '_');
              labels[lang] = norm;
            }
          }
          if (labels.isNotEmpty) {
            allProperties[pid] = labels;
          }
        }
      }
    } catch (_) {}

    processed += chunk.length;
    stdout.write('\r[Wikidata Relations] Crawled $processed / $maxProperties properties... (Found ${allProperties.length} active)');
  }
  stdout.write('\n');
  client.close();
  return allProperties;
}

void packOntologyBinary(List<(String, String, String)> entries, File outFile) {
  outFile.parent.createSync(recursive: true);

  final builder = BytesBuilder();

  // Header: Magic(8) + Version(4) + NumEntries(4) + Dim(4)
  builder.add(ascii.encode('CTRDONTO'));

  final headerData = ByteData(12);
  headerData.setUint32(0, 1, Endian.little); // version = 1
  headerData.setUint32(4, entries.length, Endian.little); // num_entries
  headerData.setUint32(8, 384, Endian.little); // dim = 384
  builder.add(headerData.buffer.asUint8List());

  for (final (lang, cat, phrase) in entries) {
    // Lang: 4 bytes fixed
    final langBytes = Uint8List(4);
    final encLang = utf8.encode(lang);
    for (var i = 0; i < encLang.length && i < 4; i++) {
      langBytes[i] = encLang[i];
    }
    builder.add(langBytes);

    // Category: 12 bytes fixed
    final catBytes = Uint8List(12);
    final encCat = utf8.encode(cat);
    for (var i = 0; i < encCat.length && i < 12; i++) {
      catBytes[i] = encCat[i];
    }
    builder.add(catBytes);

    // Text: u16 length + UTF-8 payload
    final textBytes = utf8.encode(phrase);
    final lenData = ByteData(2);
    lenData.setUint16(0, textBytes.length, Endian.little);
    builder.add(lenData.buffer.asUint8List());
    builder.add(textBytes);
  }

  final bytes = builder.takeBytes();
  outFile.writeAsBytesSync(bytes);
  print('Compiled ${entries.length} unique ontology entries into ${outFile.path} (${(bytes.length / 1024).toStringAsFixed(1)} KB)');
}

Future<void> main(List<String> args) async {
  var maxRelations = 3000;
  var maxConcepts = 10000;
  var maxSpelling = 40000;

  final relIdx = args.indexOf('--max-relations');
  if (relIdx != -1 && relIdx + 1 < args.length) {
    maxRelations = int.tryParse(args[relIdx + 1]) ?? 3000;
  }

  final conceptIdx = args.indexOf('--max-concepts');
  if (conceptIdx != -1 && conceptIdx + 1 < args.length) {
    maxConcepts = int.tryParse(args[conceptIdx + 1]) ?? 10000;
  }

  final spellIdx = args.indexOf('--spelling-words');
  if (spellIdx != -1 && spellIdx + 1 < args.length) {
    maxSpelling = int.tryParse(args[spellIdx + 1]) ?? 40000;
  }

  String? proxyArg;
  final proxyIdx = args.indexOf('--proxy');
  if (proxyIdx != -1 && proxyIdx + 1 < args.length) {
    proxyArg = args[proxyIdx + 1];
  }

  print('=' * 85);
  print('Centrode 3-Tier Multilingual Knowledge & Spelling Compiler');
  print('  • Tier 1 (Relations): $maxRelations Wikidata Properties + ConceptNet/PDTB');
  print('  • Tier 2 (Concepts):  Top $maxConcepts Core Semantic Concepts / Lang');
  print('  • Tier 3 (Spelling):  Top $maxSpelling SCOWL/Frequency Spelling Words / Lang');
  print('  • Languages:         English (en), Persian (fa), Spanish (es), Arabic (ar), Chinese (zh)');
  print('=' * 85);

  final dedupSet = <(String, String, String)>{};

  // 1. Core High-Frequency Relation Predicates & Causal Connectors
  for (final entry in coreRelationTaxonomy.entries) {
    final lang = entry.key;
    for (final p in entry.value) {
      dedupSet.add((lang, 'relation', p));
    }
  }

  // 2. Full Wikidata Relational Properties (Relations)
  final wikidataProperties = await fetchWikidataProperties(
    maxProperties: maxRelations,
    proxyAddress: proxyArg,
  );
  for (final labels in wikidataProperties.values) {
    for (final lEntry in labels.entries) {
      final lang = lEntry.key;
      final text = lEntry.value;
      if (['en', 'fa', 'es', 'ar', 'zh'].contains(lang) && text.isNotEmpty) {
        dedupSet.add((lang, 'relation', text));
      }
    }
  }

  // 3. Official Frequency Corpus: Partitioned into Tier 2 (Concepts) & Tier 3 (Spelling)
  for (final entry in languageFrequencyUrls.entries) {
    final lang = entry.key;
    final url = entry.value;
    final corpusWords = await fetchFrequencyWords(
      lang: lang,
      url: url,
      maxWords: maxSpelling,
      proxyAddress: proxyArg,
    );

    for (var i = 0; i < corpusWords.length; i++) {
      final word = corpusWords[i];
      if (i < maxConcepts) {
        dedupSet.add((lang, 'concept', word));
      }
      dedupSet.add((lang, 'spelling', word));
    }
  }

  final entries = dedupSet.toList()..sort((a, b) => a.$3.compareTo(b.$3));

  final relationCount = entries.where((e) => e.$2 == 'relation').length;
  final conceptCount = entries.where((e) => e.$2 == 'concept').length;
  final spellingCount = entries.where((e) => e.$2 == 'spelling').length;

  print('=' * 85);
  print('Final 3-Tier Ontology Statistics:');
  print('  • Tier 1 Relations (Vectors):  $relationCount entries');
  print('  • Tier 2 Concepts (Vectors):   $conceptCount entries');
  print('  • Tier 3 Spelling (Pure Text): $spellingCount entries');
  print('  • Combined Total:              ${entries.length} entries');
  print('=' * 85);

  final rootDir = Directory.current;
  final outFile = File('${rootDir.path}/assets/models/multilingual_5lang/knowledge_lexicon.bin');

  packOntologyBinary(entries, outFile);
  print('3-Tier Database & Spelling Lexicon (knowledge_lexicon.bin) Compiled Successfully!');
}
